File: mini-markov-go/markov/text.go
package markov

import (
	"math/rand"
	"regexp"
	"strings"
	"unicode"
)

// TextGenerator generates text using a Markov chain.
type TextGenerator struct {
	Chain        *Chain
	EndTokens    map[string]bool
	PreserveCase bool
}

// NewTextGenerator creates a new text generator with the specified n-gram size.
func NewTextGenerator(n int) *TextGenerator {
	return &TextGenerator{
		Chain: NewChain(n),
		EndTokens: map[string]bool{
			".": true,
			"!": true,
			"?": true,
		},
		PreserveCase: false,
	}
}

// tokenize splits text into tokens.
func (g *TextGenerator) tokenize(text string) []string {
	if !g.PreserveCase {
		text = strings.ToLower(text)
	}

	// Split on whitespace and handle punctuation
	words := strings.Fields(text)
	var tokens []string

	for _, word := range words {
		parts := splitPunctuation(word)
		for _, part := range parts {
			if part != "" {
				tokens = append(tokens, part)
			}
		}
	}

	return tokens
}

// splitPunctuation splits a word into parts with punctuation separated.
func splitPunctuation(word string) []string {
	var result []string
	var current strings.Builder

	runes := []rune(word)
	for i, r := range runes {
		if unicode.IsLetter(r) || unicode.IsDigit(r) {
			current.WriteRune(r)
		} else {
			if current.Len() > 0 {
				result = append(result, current.String())
				current.Reset()
			}
			result = append(result, string(r))
		}

		// Handle end of word
		if i == len(runes)-1 && current.Len() > 0 {
			result = append(result, current.String())
		}
	}

	return result
}

// Train trains the generator on a text corpus.
func (g *TextGenerator) Train(text string) {
	tokens := g.tokenize(text)
	if len(tokens) > 0 {
		g.Chain.Train(tokens)
	}
}

// TrainMany trains on multiple texts.
func (g *TextGenerator) TrainMany(texts []string) {
	for _, text := range texts {
		g.Train(text)
	}
}

// Generate generates text with a maximum number of words.
func (g *TextGenerator) Generate(maxWords int, rng *rand.Rand) string {
	tokens := g.Chain.Generate(maxWords, rng)
	return g.joinTokens(tokens)
}

// GenerateFrom generates text starting with a specific prompt.
func (g *TextGenerator) GenerateFrom(prompt string, maxWords int, rng *rand.Rand) string {
	promptTokens := g.tokenize(prompt)

	if len(promptTokens) < g.Chain.Order {
		return g.Generate(maxWords, rng)
	}

	result := make([]string, len(promptTokens))
	copy(result, promptTokens)

	for len(result) < len(promptTokens)+maxWords {
		current := result[len(result)-g.Chain.Order:]
		next, ok := g.Chain.SampleNext(current, rng)
		if !ok {
			break
		}
		result = append(result, next)
		if g.EndTokens[next] {
			break
		}
	}

	return g.joinTokens(result)
}

// GenerateSentence generates a complete sentence.
func (g *TextGenerator) GenerateSentence(maxWords int, rng *rand.Rand) string {
	start, ok := g.Chain.SampleStart(rng)
	if !ok {
		return ""
	}

	result := make([]string, len(start))
	copy(result, start)

	for len(result) < maxWords {
		current := result[len(result)-g.Chain.Order:]
		next, ok := g.Chain.SampleNext(current, rng)
		if !ok {
			break
		}
		result = append(result, next)
		if g.EndTokens[next] {
			break
		}
	}

	return g.joinTokens(result)
}

// joinTokens joins tokens back into text with proper spacing.
func (g *TextGenerator) joinTokens(tokens []string) string {
	if len(tokens) == 0 {
		return ""
	}

	punctuation := regexp.MustCompile(`^[.!?,;:'")\]]$`)
	var result strings.Builder

	for i, token := range tokens {
		if i > 0 && !punctuation.MatchString(token) {
			result.WriteString(" ")
		}
		result.WriteString(token)
	}

	return result.String()
}

// Stats returns statistics about the generator.
func (g *TextGenerator) Stats() TextGeneratorStats {
	return TextGeneratorStats{
		Order:          g.Chain.Order,
		NumStates:      g.Chain.NumStates(),
		NumTransitions: g.Chain.NumTransitions(),
		NumSequences:   g.Chain.NumSequences,
		Entropy:        g.Chain.Entropy(),
	}
}

// TextGeneratorStats holds statistics about a text generator.
type TextGeneratorStats struct {
	Order          int
	NumStates      int
	NumTransitions int
	NumSequences   int
	Entropy        float64
}

File: mini-markov-python/mini_markov/text.py
"""Text generation using Markov chains."""

from typing import List, Set, Optional
import re
import random

from .chain import MarkovChain


class TextGenerator:
    """
    Text generator using word-level Markov chains.
    """
    
    def __init__(self, n: int = 2, preserve_case: bool = False):
        """
        Creates a new text generator.
        
        Args:
            n: The n-gram size (1 = unigram, 2 = bigram, etc.).
            preserve_case: Whether to preserve letter case.
        """
        self.chain = MarkovChain[str](n)
        self.preserve_case = preserve_case
        self.end_tokens: Set[str] = {".", "!", "?"}
    
    def add_end_token(self, token: str) -> "TextGenerator":
        """Adds a custom end token."""
        self.end_tokens.add(token)
        return self
    
    def _tokenize(self, text: str) -> List[str]:
        """Tokenizes text into words and punctuation."""
        if not self.preserve_case:
            text = text.lower()
        
        # Split on whitespace first
        tokens = []
        for word in text.split():
            tokens.extend(self._split_punctuation(word))
        
        return [t for t in tokens if t]
    
    def _split_punctuation(self, word: str) -> List[str]:
        """Splits punctuation from words."""
        result = []
        current = []
        
        for char in word:
            if char.isalnum():
                current.append(char)
            else:
                if current:
                    result.append("".join(current))
                    current = []
                result.append(char)
        
        if current:
            result.append("".join(current))
        
        return result
    
    def train(self, text: str) -> None:
        """Trains the generator on a text corpus."""
        tokens = self._tokenize(text)
        if tokens:
            self.chain.train(tokens)
    
    def train_many(self, *texts: str) -> None:
        """Trains on multiple texts."""
        for text in texts:
            self.train(text)
    
    def generate(self, max_words: int, rng: random.Random = None) -> str:
        """Generates text with a maximum number of words."""
        tokens = self.chain.generate(max_words, rng)
        return self._join_tokens(tokens)
    
    def generate_from(self, prompt: str, max_words: int, rng: random.Random = None) -> str:
        """Generates text starting with a specific prompt."""
        prompt_tokens = self._tokenize(prompt)
        
        if len(prompt_tokens) < self.chain.order:
            return self.generate(max_words, rng)
        
        result = list(prompt_tokens)
        
        while len(result) < len(prompt_tokens) + max_words:
            current = result[-self.chain.order:]
            next_token = self.chain.sample_next(current, rng)
            if next_token is None:
                break
            result.append(next_token)
            if next_token in self.end_tokens:
                break
        
        return self._join_tokens(result)
    
    def generate_sentence(self, max_words: int, rng: random.Random = None) -> str:
        """Generates a complete sentence."""
        start = self.chain.sample_start(rng)
        if start is None:
            return ""
        
        result = list(start)
        
        while len(result) < max_words:
            current = result[-self.chain.order:]
            next_token = self.chain.sample_next(current, rng)
            if next_token is None:
                break
            result.append(next_token)
            if next_token in self.end_tokens:
                break
        
        return self._join_tokens(result)
    
    def _join_tokens(self, tokens: List[str]) -> str:
        """Joins tokens back into text with proper spacing."""
        if not tokens:
            return ""
        
        punctuation = re.compile(r'^[.!?,;:\'")\]]$')
        result = []
        
        for i, token in enumerate(tokens):
            if i > 0 and not punctuation.match(token):
                result.append(" ")
            result.append(token)
        
        return "".join(result)
    
    @property
    def stats(self) -> dict:
        """Returns statistics about the generator."""
        return {
            "order": self.chain.order,
            "num_states": self.chain.num_states,
            "num_transitions": self.chain.num_transitions,
            "num_sequences": self.chain.num_sequences,
            "entropy": self.chain.entropy(),
        }
    
    def __repr__(self) -> str:
        return f"TextGenerator(order={self.chain.order}, states={self.chain.num_states})"


class CharGenerator:
    """
    Character-level text generator using Markov chains.
    """
    
    def __init__(self, n: int = 3):
        """
        Creates a new character generator.
        
        Args:
            n: The n-gram size (number of characters to consider).
        """
        self.chain = MarkovChain[str](n)
    
    def train(self, text: str) -> None:
        """Trains the generator on text."""
        chars = list(text)
        if chars:
            self.chain.train(chars)
    
    def generate(self, max_chars: int, rng: random.Random = None) -> str:
        """Generates text with a maximum number of characters."""
        chars = self.chain.generate(max_chars, rng)
        return "".join(chars)
    
    @property
    def stats(self) -> dict:
        """Returns statistics about the generator."""
        return {
            "order": self.chain.order,
            "num_states": self.chain.num_states,
            "num_transitions": self.chain.num_transitions,
            "entropy": self.chain.entropy(),
        }
    
    def __repr__(self) -> str:
        return f"CharGenerator(order={self.chain.order}, states={self.chain.num_states})"
